src/AWS/Orleans.Reminders.DynamoDB/Reminders/DynamoDBReminderTable.cs

```diff
@@ -146,31 +146,48 @@ public async Task<ReminderTableData> ReadRows(GrainReference grainRef)
-        /// <param name="endHash"></param>
+        /// <param name="begin"></param>
+        /// <param name="end"></param>
         /// <returns> Return the RemiderTableData if the rows were read successfully </returns>
-        public async Task<ReminderTableData> ReadRows(uint beginHash, uint endHash)
+        public async Task<ReminderTableData> ReadRows(uint begin, uint end)
         {
-            var expressionValues = new Dictionary<string, AttributeValue>
-                {
-                    { $":{SERVICE_ID_PROPERTY_NAME}", new AttributeValue(this.serviceId) },
-                    { $":Begin{GRAIN_HASH_PROPERTY_NAME}", new AttributeValue { N = beginHash.ToString() } },
-                    { $":End{GRAIN_HASH_PROPERTY_NAME}", new AttributeValue { N = endHash.ToString() } }
-                };
+            Dictionary<string, AttributeValue> expressionValues = null;
 
             try
             {
                 string expression = string.Empty;
-                if (beginHash < endHash)
+                List<ReminderEntry> records;
+
+                if (begin < end)
                 {
-                    expression = $"{SERVICE_ID_PROPERTY_NAME} = :{SERVICE_ID_PROPERTY_NAME} AND {GRAIN_HASH_PROPERTY_NAME} > :Begin{GRAIN_HASH_PROPERTY_NAME} AND {GRAIN_HASH_PROPERTY_NAME} <= :End{GRAIN_HASH_PROPERTY_NAME}";
+                    expressionValues = new Dictionary<string, AttributeValue>
+                    {
+                        { $":{SERVICE_ID_PROPERTY_NAME}", new AttributeValue(this.serviceId) },
+                        { $":Begin{GRAIN_HASH_PROPERTY_NAME}", new AttributeValue { N = begin.ToString() } },
+                        { $":End{GRAIN_HASH_PROPERTY_NAME}", new AttributeValue { N = end.ToString() } }
+                    };
+                    expression = $"{SERVICE_ID_PROPERTY_NAME} = :{SERVICE_ID_PROPERTY_NAME} AND {GRAIN_HASH_PROPERTY_NAME} BETWEEN :Begin{GRAIN_HASH_PROPERTY_NAME} AND :End{GRAIN_HASH_PROPERTY_NAME}";
+                    records = await this.storage.QueryAllAsync(this.options.TableName, expressionValues, expression, this.Resolve, SERVICE_ID_INDEX, consistentRead: false).ConfigureAwait(false);
                 }
                 else
                 {
-                    expression = $"{SERVICE_ID_PROPERTY_NAME} = :{SERVICE_ID_PROPERTY_NAME} AND ({GRAIN_HASH_PROPERTY_NAME} > :Begin{GRAIN_HASH_PROPERTY_NAME} OR {GRAIN_HASH_PROPERTY_NAME} <= :End{GRAIN_HASH_PROPERTY_NAME})";
-                }
+                    expressionValues = new Dictionary<string, AttributeValue>
+                    {
+                        { $":{SERVICE_ID_PROPERTY_NAME}", new AttributeValue(this.serviceId) },
+                        { $":End{GRAIN_HASH_PROPERTY_NAME}", new AttributeValue { N = end.ToString() } }
+                    };
+                    expression = $"{SERVICE_ID_PROPERTY_NAME} = :{SERVICE_ID_PROPERTY_NAME} AND {GRAIN_HASH_PROPERTY_NAME} <= :End{GRAIN_HASH_PROPERTY_NAME}";
```

Comment:
This query should be `<= End || > Begin` (since "end" is smaller than "begin", it's a wrap-around query).

Here's a graphical depiction of the two queries, where a circle represents an exclusive bound and a bar represents an inclusive bound:

![image](https://user-images.githubusercontent.com/203839/148422757-334a8a5d-3060-4006-babd-31bbf5d04a37.png)

Reply:
Indeed, the original query had this right: `> Begin or <= End`: it's a conjunction of two queries

Reply:
Originally, this was performing a scan.  I don't believe you can perform a query with the "or" clause, and thus I split the scan into two queries, then combine the results.

Reply:
Oh, I see now. That's an odd limitation, but nevertheless this seems fine.